File: src/main/java/it/giaquinto/springberry/model/user/implementation/BaseUser.java
package it.giaquinto.springberry.model.user.implementation;

import it.giaquinto.springberry.model.user.User;

import java.time.LocalDateTime;
import java.util.Objects;

public abstract class BaseUser implements User {

    private final String name;
    private final String surname;

    private final LocalDateTime created;

    public BaseUser(
            final String name,
            final String surname,
            final LocalDateTime created
    ) {
        this.name = name;
        this.surname = surname;
        this.created = created;
    }

    public String getName() {
        return name;
    }

    public String getSurname() {
        return surname;
    }

    public LocalDateTime getCreated() {
        return created;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (!(o instanceof BaseUser baseUser)) return false;
        return Objects.equals(name, baseUser.name) && Objects.equals(surname, baseUser.surname) && Objects.equals(created, baseUser.created);
    }

    @Override
    public int hashCode() {
        return Objects.hash(name, surname, created);
    }
}

File: src/main/java/it/giaquinto/springberry/model/user/implementation/BaseUser.kt
package it.giaquinto.springberry.model.user.implementation

import it.giaquinto.springberry.model.user.User
import java.time.LocalDateTime
import java.util.*

abstract class BaseUser(
    val name: String?,
    val surname: String?,
    val created: LocalDateTime?
) : User {

    override fun equals(o: Any?): Boolean {
        if (this === o) return true
        return if (o !is BaseUser) false else name == o.name && surname == o.surname && created == o.created
    }

    override fun hashCode(): Int {
        return Objects.hash(name, surname, created)
    }
}